// rsbuild-plugin-dts@0.21.3/dist/index.js
import { fork } from "node:child_process";
import { extname, join } from "node:path";
import { logger } from "@rsbuild/core";
import { processSourceEntry, cleanDtsFiles, clearTempDeclarationDir, getDtsEmitPath, cleanTsBuildInfoFile, color, warnIfOutside, ts, loadTsconfig } from "./350.js";
const PLUGIN_DTS_NAME = 'rsbuild:dts';
const pluginDts = (options = {})=>({
        name: PLUGIN_DTS_NAME,
        setup (api) {
            const loggerLevel = api.logger.level;
            logger.level = loggerLevel;
            let apiExtractorOptions = {};
            if (options.bundle && 'object' == typeof options.bundle) apiExtractorOptions = {
                ...options.bundle
            };
            const bundle = !!options.bundle;
            options.abortOnError = options.abortOnError ?? true;
            options.build = options.build ?? false;
            options.redirect = options.redirect ?? {};
            options.redirect.path = options.redirect.path ?? true;
            options.redirect.extension = options.redirect.extension ?? false;
            options.alias = options.alias ?? {};
            options.tsgo = options.tsgo ?? false;
            let dtsPromise = Promise.resolve({
                status: 'success'
            });
            let promiseResult;
            let childProcesses = [];
            api.onBeforeEnvironmentCompile(async ({ isWatch, isFirstCompile, environment })=>{
                if (!isFirstCompile && !options.tsgo) return;
                const { config } = environment;
                const dtsEntry = processSourceEntry(bundle, config.source?.entry);
                const cwd = api.context.rootPath;
                const tsconfigPath = ts.findConfigFile(cwd, ts.sys.fileExists.bind(ts.sys), config.source.tsconfigPath);
                if (!tsconfigPath) {
                    const error = new Error(`Failed to resolve tsconfig file ${color.cyan(`"${config.source.tsconfigPath}"`)} from ${color.cyan(cwd)}. Please ensure that the file exists.`);
                    error.stack = '';
                    throw error;
                }
                const tsConfigResult = loadTsconfig(tsconfigPath);
                const { options: rawCompilerOptions } = tsConfigResult;
                const { declarationDir, outDir, composite, incremental } = rawCompilerOptions;
                const distPathRoot = 'string' == typeof config.output?.distPath ? config.output?.distPath : config.output?.distPath.root;
                const dtsEmitPath = getDtsEmitPath(options.distPath, declarationDir, distPathRoot);
                warnIfOutside(cwd, declarationDir, 'declarationDir');
                warnIfOutside(cwd, outDir, 'outDir');
                if (false !== config.output.cleanDistPath) await cleanDtsFiles(cwd, dtsEmitPath);
                if (bundle) await clearTempDeclarationDir(cwd);
                if (composite || incremental || options.build) await cleanTsBuildInfoFile(tsconfigPath, rawCompilerOptions);
                const jsExtension = extname(import.meta.filename);
                const childProcess = fork(join(import.meta.dirname, `./dts${jsExtension}`), [], {
                    stdio: 'inherit'
                });
                childProcesses.push(childProcess);
                const dtsGenOptions = {
                    ...options,
                    bundle,
                    dtsEntry,
                    dtsEmitPath,
                    userExternals: config.output.externals,
                    apiExtractorOptions,
                    tsconfigPath,
                    tsConfigResult,
                    name: environment.name,
                    cwd,
                    isWatch,
                    loggerLevel: loggerLevel
                };
                childProcess.send(dtsGenOptions);
                dtsPromise = new Promise((resolve)=>{
                    childProcess.on('message', (message)=>{
                        if ('success' === message) resolve({
                            status: 'success'
                        });
                        else if ('error' === message) resolve({
                            status: 'error',
                            errorMessage: `Error occurred in ${environment.name} declaration files generation.`
                        });
                    });
                });
            });
            api.onAfterBuild({
                handler: async ({ isFirstCompile })=>{
                    if (!isFirstCompile) return;
                    promiseResult = await dtsPromise;
                },
                order: 'pre'
            });
            api.onAfterBuild(({ isFirstCompile })=>{
                if (!isFirstCompile) return;
                if ('error' === promiseResult.status) {
                    if (options.abortOnError) {
                        const error = new Error(promiseResult.errorMessage);
                        error.stack = '';
                        throw error;
                    }
                    promiseResult.errorMessage && logger.error(promiseResult.errorMessage);
                    logger.warn('With `abortOnError` configuration currently disabled, type errors will not fail the build, but proper type declaration output cannot be guaranteed.');
                }
            });
            const killProcesses = ()=>{
                for (const childProcess of childProcesses)if (!childProcess.killed) try {
                    childProcess.kill();
                } catch  {}
                childProcesses = [];
            };
            api.onCloseBuild(killProcesses);
            api.onCloseDevServer(killProcesses);
        }
    });
export { PLUGIN_DTS_NAME, pluginDts };
